Fix-#3296: Patch __array__ method for RPyC proxy objects

Lozovskii-Aleksandr opened the pull request:
<!--
Thank you for your contribution!
Please review the contributing docs: https://modin.readthedocs.io/en/latest/CONTRIBUTING.html
if you have questions about contributing.
-->

## What do these changes do?

<!-- Please give a short brief about these changes. -->

- [X] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [X] passes `flake8 modin`
- [X] passes `black --check modin`
- [X] signed commit with `git commit -s` <!-- you can amend your commit with a signature via `git commit -amend -s` -->
- [X] Resolves #3296? 
- [X] tests added and passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date <!-- if you have added, renamed or removed files or directories please update the documentation accordingly -->

## Solution description

The error was in the module `rpyc.core.netref`. 

 For the solution, it was necessary to patch the `_make_method` method. In the `_make_method` method, in the `name == "__array__ "` block, the `_array__` method must accept two parameters: `self` and `type`. Adding logic from `rpyc_patches.py`, I am correcting this error. Now the `__array__` method accepts the `type` parameter and processes it correctly. More information about the solution is written in https://github.com/modin-project/modin/issues/3296.

vnlitvinov commented on modin/experimental/cloud/rpyc_patches.py:
```diff
@@ -0,0 +1,26 @@
+                )
+
+                if dtype is not None:
+                    import numpy
```
I propose to move that `import` to the top of the file

vnlitvinov reviewed (commented):
Please fix the CI and clean up commit history a little bit.

vnlitvinov commented on .github/workflows/ci.yml:
```diff
@@ -547,6 +547,9 @@ jobs:
       - shell: bash -l {0}
         run: python -m pytest --simulate-cloud=normal modin/pandas/test/test_io.py::TestCsv
+      - shell: bash -l {0}
+      # When running without parameters, some of the tests fall
```
```suggestion
      # When running without parameters, some of the tests fail
```

vnlitvinov commented on modin/experimental/cloud/rpyc_patches.py:
```diff
@@ -0,0 +1,25 @@
+                # Note that protocol=-1 will only work between python
+                # interpreters of the same version.
+                res = netref.pickle.loads(
+                    netref.syncreq(self, netref.consts.HANDLE_PICKLE, -1)
```
```suggestion
                    netref.syncreq(self, netref.consts.HANDLE_PICKLE, pickle.HIGHEST_PROTOCOL)
```
let's be explicit in protocol versioning to remove this uncertainty about Python versions' compatibility

vnlitvinov reviewed (commented):
Changes look good, let's wait until CI passes.

vnlitvinov reviewed (approved):
LGTM when tests pass.

YarShev commented:
@Lozovskii-Aleksandr , why wasn't any test added to CI to verify the issue resolved?

Lozovskii-Aleksandr commented:
> @Lozovskii-Aleksandr , why wasn't any test added to CI to verify the issue resolved?

@YarShev, I added one test to `ci.yml` that didn't work before, and after fixing it, it was fixed. It would be possible to add a group of tests that do not have parameters, but it will take a very long time to run. If I'm not mistaken, there are ~40 tests running for 20+ minutes. It would be very difficult to find something faster, since many tests have already failed due to other errors. Therefore, I added only one test.

YarShev commented:
@Lozovskii-Aleksandr , Shouldn't we also add the tests you mentioned in the issue to CI. I mean these ones:

* modin/pandas/test/dataframe/test_join_sort.py::test_sort_values[None-ignore_index_True-first-quicksort-inplace_True-ascending_True-over_rows_int-multiindex_level0-int_data]
* modin/pandas/test/dataframe/test_join_sort.py::test_sort_values[None-ignore_index_True-first-quicksort-inplace_True-ascending_True-over_rows_int-multiindex_level1,multiindex_level0-int_data]
* modin/pandas/test/dataframe/test_join_sort.py::test_sort_values[None-ignore_index_True-first-quicksort-inplace_True-ascending_True-over_rows_int-multiindex_level0,last,first,multiindex_level1-int_data]
* modin/pandas/test/dataframe/test_indexing.py::test_reset_index_with_multi_index_no_drop[True-0-False-no_col_fill-no_col_level-no_level-True-3-float_nan_data]
* modin/pandas/test/dataframe/test_indexing.py::test_reset_index_with_multi_index_no_drop[True-0-False-no_col_fill-no_col_level-None-True-3-float_nan_data]
* modin/pandas/test/dataframe/test_indexing.py::test_reset_index_with_multi_index_no_drop[True-0-False-no_col_fill-no_col_level-0-True-3-float_nan_data]

It is said in the issue, your changes might fix the two problems that are similar but not identical.

Lozovskii-Aleksandr commented:
@YarShev, The tests listed by you are not passing now. There are still errors in them that will need to be corrected. But the error that was corrected in this PR is no longer found in them. 

In the problem, I said that the first and second problems are as similar as possible, and the third is probably related to them. This turned out to be true, since after the correction, the specified error ceased to occur in them.

YarShev commented:
@Lozovskii-Aleksandr , ok, got it!
